### PR TITLE
[SPARK-40019][SQL] Refactor comment of ArrayType's containsNull and refactor the misunderstanding logics in collectionOperator's expression about `containsNull`

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
ArrayType's parameter `containsNull` means this array can contains null, related to nullable, this is easy to misunderstand in reading logic. In this pr, we  refactor the comment about `containsNull` and refactor the code in ArrayType related expression to make the code path have a certain meaning.

### Why are the changes needed?
Refactor code

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Not need

### CODE REVIEW

**Review by HyukjinKwon — COMMENTED** [review]
I am -1 on this change. it's confusing if this indicates `array(null)` or `null`. In addition, this is related to the public API, e.g., `ArrayType`.

In addition, there's `MapType.valueContainsNull` too.

**Comment by AngersZhuuuu** [comment]
> I am -1 on this change. it's confusing if this indicates `array(null)` or `null`. In addition, this is related to the public API, e.g., `ArrayType`.
> 
> In addition, there's `MapType.valueContainsNull` too.

Have discussed with wenchen, change this later

**Comment by AngersZhuuuu** [comment]
> I am -1 on this change. it's confusing if this indicates `array(null)` or `null`. In addition, this is related to the public API, e.g., `ArrayType`.
> 
> In addition, there's `MapType.valueContainsNull` too.

How about the current? I know MapType have the same problem and I will raise a pr for MapType too

**Comment by HyukjinKwon** [comment]
Yeah, I am fine with the current approach.

**Comment by AngersZhuuuu** [comment]
> Yeah, I am fine with the current approach.

Any more suggestions? also ping @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -1079,6 +1084,9 @@ case class Shuffle(child: Expression, randomSeed: Option[Long] = None)
 
   override def dataType: DataType = child.dataType
 
+  @transient private lazy val resultArrayElementNullable =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1087]:
```suggestion
  private def resultArrayElementNullable =
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -1162,6 +1170,9 @@ case class Reverse(child: Expression)
...
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:1173]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/ArrayType.scala [hunk 3]
```diff
@@ -53,11 +53,12 @@ object ArrayType extends AbstractDataType {
...
  * An [[ArrayType]] object comprises two fields, `elementType: [[DataType]]` and
- * `containsNull: Boolean`. The field of `elementType` is used to specify the type of
- * array elements. The field of `containsNull` is used to specify if the array has `null` values.
+ * `containsNull: Boolean`.
+ * The field of `elementType` is used to specify the type of array elements.
+ * The field of `containsNull` is used to specify if the array can have `null` values.
  *
  * @param elementType The data type of values.
- * @param containsNull Indicates if values have `null` values
+ * @param containsNull Indicates if values can have `null` values
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/ArrayType.scala:61]:
```suggestion
 * @param containsNull Indicates if the array can have `null` values
```

**Comment by cloud-fan** [comment]
can you update the PR title?

**Comment by AngersZhuuuu** [comment]
> can you update the PR title?

How  about current?

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Pull request closed without merging**